code/modules/projectiles/guns/energy.dm

```diff
@@ -594,3 +594,776 @@
+	name = "\improper TE-ML Machine laser rifle"
+	desc = "A TerraGov standard issue sniper rifle with an integrated charge selector for normal, high and heat settings. Uses standard TE cells."
+	reload_sound = 'sound/weapons/guns/interact/TE-M Reload.ogg'
+	fire_sound = 'sound/weapons/guns/fire/Laser Rifle Standard.ogg'
+	force = 20
+	icon_state = "tem"
+	item_state = "tem"
+	icon = 'icons/Marine/gun64.dmi'
+	w_class = WEIGHT_CLASS_BULKY
+	max_shots = 200 //codex stuff
+	load_method = CELL //codex stuff
+	ammo = /datum/ammo/energy/lasgun/marine/autolaser
+	ammo_diff = null
+	cell_type = /obj/item/cell/lasgun/marine
+	charge_cost = 3
+	gun_firemode = GUN_FIREMODE_AUTOMATIC
+	gun_firemode_list = list(GUN_FIREMODE_SEMIAUTO, GUN_FIREMODE_AUTOMATIC)
+
+	attachable_allowed = list(
+		/obj/item/attachable/bayonet,
+		/obj/item/attachable/bayonetknife,
+		/obj/item/attachable/reddot,
+		/obj/item/attachable/verticalgrip,
+		/obj/item/attachable/angledgrip,
+		/obj/item/attachable/lasersight,
+		/obj/item/attachable/gyro,
+		/obj/item/attachable/flashlight,
+		/obj/item/attachable/bipod,
+		/obj/item/attachable/magnetic_harness,
+		/obj/item/attachable/attached_gun/grenade,
+		/obj/item/attachable/scope,
+		/obj/item/attachable/scope/mini,
+	)
+
+	flags_gun_features = GUN_AUTO_EJECTOR|GUN_CAN_POINTBLANK|GUN_AMMO_COUNTER|GUN_ENERGY|GUN_AMMO_COUNTER
+	actions_types = list(/datum/action/item_action/aim_mode)
+	attachable_offset = list("muzzle_x" = 32, "muzzle_y" = 18,"rail_x" = 12, "rail_y" = 23, "under_x" = 23, "under_y" = 15, "stock_x" = 22, "stock_y" = 12)
+
+	aim_slowdown = 1
+	wield_delay = 1.2 SECONDS
+	scatter = 0
+	scatter_unwielded = 10
+	aim_fire_delay = 0.1 SECONDS
+	fire_delay = 0.2 SECONDS
+	accuracy_mult = -0.1
+	accuracy_mult_unwielded = 0.6
+	scatter_unwielded = 10
+	var/list/datum/energy_ml_mode/mode_list = list(
+		"Standard" = /datum/energy_ml_mode/standard,
+	)
+
+
+/datum/energy_ml_mode
+	///how much power the gun uses on this mode when shot.
+	var/charge_cost = 0
+	///the ammo datum this mode is.
+	var/ammo = null
+	///how long it takes between each shot of that mode, same as gun fire delay.
+	var/fire_delay = null
+	///The gun firing sound of this mode.
+	var/fire_sound = 0
+	///What message it sends to the user when you switch to this mode.
+	var/message_to_user = ""
+	///Used to change the gun firemode, like automatic, semi-automatic and burst.
+	var/fire_mode = GUN_FIREMODE_SEMIAUTO
+	///what to change the gun icon_state to when switching to this mode.
+	var/icon_state = "ter"
+	///Which icon file the radial menu will use.
+	var/radial_icon = 'icons/mob/radial.dmi'
+	///The icon state the radial menu will use.
+	var/radial_icon_state = "laser"
+
+/datum/energy_ml_mode/standard
+	charge_cost = 3
+	ammo = /datum/ammo/energy/lasgun/marine/autolaser
+	fire_delay = 0.2 SECONDS
+	fire_sound = 'sound/weapons/guns/fire/Laser Rifle Standard.ogg'
+	message_to_user = "You set the Laser rifle's charge mode to standard fire."
```

Comment:
```suggestion
	message_to_user = "You set the [src]'s charge mode to standard fire."
```
Also, because there are no charges modes, do you think it would be neccessary?

Reply:
Charge modes exist on all guns expect for the LM which might get eco fire later